segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -136,52 +146,174 @@ public void initialize(long epoch) {
+                            HDFSMetrics.READ_LATENCY.reportSuccessEvent(timer.getElapsed());
+                            HDFSMetrics.READ_BYTES.add(totalBytesRead);
+                            LoggerHelpers.traceLeave(log, "read", traceId, handle, offset, totalBytesRead);
+                            return totalBytesRead;
+                        });
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.throwException(handle.getSegmentName(), e);
+        } catch (RetriesExhaustedException e) {
+            throw HDFSExceptionHelpers.throwException(handle.getSegmentName(), e.getCause());
+        }
     }
 
     @Override
-    public void write(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentException {
-        run(new WriteOperation(asWritableHandle(handle), offset, data, length, this.context));
+    public SegmentHandle openRead(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+        try {
+            //Ensure that file exists
+            findStatusForSegment(streamSegmentName, true);
+            LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+            return HDFSSegmentHandle.read(streamSegmentName);
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.throwException(streamSegmentName, e);
+        }
     }
 
     @Override
     public void seal(SegmentHandle handle) throws StreamSegmentException {
-        run(new SealOperation(asWritableHandle(handle), this.context));
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle);
+        handle = asWritableHandle(handle);
+        try {
+            FileStatus lastHandleFile = findStatusForSegment(handle.getSegmentName(), true);
+
+            if (!isSealed(lastHandleFile.getPath())) {
+                if (getEpoc(lastHandleFile) > this.epoch) {
+                    throw new StorageNotPrimaryException(handle.getSegmentName());
+                }
+                makeReadOnly(lastHandleFile);
+                Path sealedPath = getSealedFilePath(handle.getSegmentName());
+                this.fileSystem.rename(lastHandleFile.getPath(), sealedPath);
+            }
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.throwException(handle.getSegmentName(), e);
+        }
+        LoggerHelpers.traceLeave(log, "seal", traceId, handle);
     }
 
     @Override
     public void unseal(SegmentHandle handle) throws StreamSegmentException {
-        run(new UnsealOperation(asReadableHandle(handle), this.context));
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle);
+        try {
+            FileStatus status = findStatusForSegment(handle.getSegmentName(), true);
+            if (!isSealed(status.getPath())) {
+                throw new StorageNotPrimaryException(handle.getSegmentName());
```

Comment:
Why throw this kind of exception if the file is not sealed?

Reply:
`SyncStorage.unseal` says to throw this exception when this Storage instance is no longer primary for this segment. It does not say anything about whether the file is already sealed.